IDE0004 Unnecessary cast false positive

**Version Used**: 

main

**Steps to Reproduce**:

```csharp
class C
{
    static object M(object o)
    {
        return (object)o.ToString;
    }
}
```

**Expected Behavior**:

No unnecessary cast is reported.

**Actual Behavior**:

Unnecessary cast is reported. Removing the cast will produce this warning:

> warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?

Extracted from https://github.com/dotnet/roslyn/issues/50043#issuecomment-924209443

--- a/src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs
+++ b/src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs
@@ -12647,7 +12647,7 @@ class C
 
         [WorkItem(58804, "https://github.com/dotnet/roslyn/issues/58804")]
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryCast)]
-        public async Task ConvertingMethodGroupToObject()
+        public async Task ConvertingMethodGroupToObject_CastIsNecessary()
         {
             var code = @"
 class C
@@ -12663,6 +12663,45 @@ class C
                 TestCode = code,
                 FixedCode = code,
                 LanguageVersion = LanguageVersion.CSharp10,
+            }.RunAsync();
+        }
+
+        [WorkItem(58804, "https://github.com/dotnet/roslyn/issues/58804")]
+        [Theory, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryCast)]
+        [InlineData("Delegate")]
+        [InlineData("MulticastDelegate")]
+        [InlineData("Func<string>")]
+        public async Task ConvertingMethodGroupToObject_CastIsUnnecessary(string type)
+        {
+            var code = $@"
+using System;
+
+class C
+{{
+    static {type} M(object o)
+    {{
+        return ({type})[|(object)|]o.ToString;
+    }}
+}}
+";
+            var fixedCode = $@"
+using System;
+
+class C
+{{
+    static {type} M(object o)
+    {{
+        return o.ToString;
+    }}
+}}
+";
+            await new VerifyCS.Test
+            {
+                TestCode = code,
+                FixedCode = fixedCode,
+                LanguageVersion = LanguageVersion.CSharp10,
+                NumberOfIncrementalIterations = 2,
+                NumberOfFixAllIterations = 2,
             }.RunAsync();
         }
     }

--- a/src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs
+++ b/src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs
@@ -12644,5 +12644,26 @@ class C
                 LanguageVersion = LanguageVersion.CSharp10,
             }.RunAsync();
         }
+
+        [WorkItem(58804, "https://github.com/dotnet/roslyn/issues/58804")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryCast)]
+        public async Task ConvertingMethodGroupToObject()
+        {
+            var code = @"
+class C
+{
+    static object M(object o)
+    {
+        return (object)o.ToString;
+    }
+}
+";
+            await new VerifyCS.Test
+            {
+                TestCode = code,
+                FixedCode = code,
+                LanguageVersion = LanguageVersion.CSharp10,
+            }.RunAsync();
+        }
     }
 }

--- a/src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs
+++ b/src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs
@@ -12667,6 +12667,34 @@ class C
         }
 
         [WorkItem(58804, "https://github.com/dotnet/roslyn/issues/58804")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryCast)]
+        public async Task ConvertingMethodGroupToObject_CastIsNecessary2()
+        {
+            var code = @"
+using System;
+
+class C
+{
+    static T M<T>(object o)
+    {
+        return (T)(object)o.ToString;
+    }
+
+    static T M2<T>(object o) where T : Delegate
+    {
+        return (T)(object)o.ToString;
+    }
+}
+";
+            await new VerifyCS.Test
+            {
+                TestCode = code,
+                FixedCode = code,
+                LanguageVersion = LanguageVersion.CSharp10,
+            }.RunAsync();
+        }
+
+        [WorkItem(58804, "https://github.com/dotnet/roslyn/issues/58804")]
         [Theory, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryCast)]
         [InlineData("Delegate")]
         [InlineData("MulticastDelegate")]
@@ -12694,6 +12722,42 @@ class C
         return o.ToString;
     }}
 }}
+";
+            await new VerifyCS.Test
+            {
+                TestCode = code,
+                FixedCode = fixedCode,
+                LanguageVersion = LanguageVersion.CSharp10,
+                NumberOfIncrementalIterations = 2,
+                NumberOfFixAllIterations = 2,
+            }.RunAsync();
+        }
+
+        [WorkItem(58804, "https://github.com/dotnet/roslyn/issues/58804")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryCast)]
+        public async Task ConvertingMethodGroupToObject_CastIsUnnecessary2()
+        {
+            var code = @"
+using System;
+
+class C
+{
+    static Delegate M(object o)
+    {
+        return (Delegate)[|(object)|]o.ToString;
+    }
+}
+";
+            var fixedCode = @"
+using System;
+
+class C
+{
+    static Delegate M(object o)
+    {
+        return o.ToString;
+    }
+}
 ";
             await new VerifyCS.Test
             {

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
@@ -81,6 +81,11 @@ namespace Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers
             if (IsRemovableBitwiseEnumNegation(cast, semanticModel, cancellationToken))
                 return true;
 
+            // Special case for converting a method group to object. The compiler issues a warning if the cast is removed:
+            // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
+            if (semanticModel.GetOperation(cast.Expression)?.Kind == OperationKind.MethodReference)
+                return false;
+
             return IsCastSafeToRemove(cast, cast.Expression, semanticModel, cancellationToken);
         }
 

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
@@ -84,7 +84,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers
             // Special case for converting a method group to object. The compiler issues a warning if the cast is removed:
             // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
             if (semanticModel.GetOperation(cast.Expression, cancellationToken)?.Kind == OperationKind.MethodReference &&
-                semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type is { SpecialType: SpecialType.System_Object })
+                semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type?.SpecialType is SpecialType.System_Object)
             {
                 return false;
             }

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
@@ -83,10 +83,16 @@ namespace Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers
 
             // Special case for converting a method group to object. The compiler issues a warning if the cast is removed:
             // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
-            if (semanticModel.GetOperation(cast.Expression, cancellationToken)?.Kind == OperationKind.MethodReference &&
+            var castExpressionOperation = semanticModel.GetOperation(cast.Expression, cancellationToken);
+            if (castExpressionOperation?.Kind == OperationKind.MethodReference &&
                 semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type?.SpecialType is SpecialType.System_Object)
             {
-                return false;
+                // If we have a double cast, report as unnecessary, e.g:
+                // (object)(object)MethodGroup
+                // (Delegate)(object)MethodGroup
+                // If we have a single object cast, don't report as unnecessary e.g:
+                // (object)MethodGroup
+                return castExpressionOperation.Parent?.Parent?.Kind == OperationKind.Conversion;
             }
 
             return IsCastSafeToRemove(cast, cast.Expression, semanticModel, cancellationToken);

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
@@ -84,15 +84,25 @@ namespace Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers
             // Special case for converting a method group to object. The compiler issues a warning if the cast is removed:
             // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
             var castExpressionOperation = semanticModel.GetOperation(cast.Expression, cancellationToken);
-            if (castExpressionOperation?.Kind == OperationKind.MethodReference &&
-                semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type?.SpecialType is SpecialType.System_Object)
+            if (castExpressionOperation is
+                {
+                    Kind: OperationKind.MethodReference,
+                    Parent.Kind: OperationKind.DelegateCreation,
+                    Parent.Parent: IConversionOperation { Type.SpecialType: SpecialType.System_Object } conversionOperation
+                })
             {
                 // If we have a double cast, report as unnecessary, e.g:
                 // (object)(object)MethodGroup
                 // (Delegate)(object)MethodGroup
                 // If we have a single object cast, don't report as unnecessary e.g:
                 // (object)MethodGroup
-                return castExpressionOperation.Parent?.Parent?.Kind == OperationKind.Conversion;
+                if (conversionOperation.Parent is IConversionOperation { Type: { } parentConversionType } &&
+                    semanticModel.ClassifyConversion(cast.Expression, parentConversionType).Exists)
+                {
+                    return true;
+                }
+
+                return false;
             }
 
             return IsCastSafeToRemove(cast, cast.Expression, semanticModel, cancellationToken);

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
@@ -83,7 +83,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers
 
             // Special case for converting a method group to object. The compiler issues a warning if the cast is removed:
             // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
-            if (semanticModel.GetOperation(cast.Expression)?.Kind == OperationKind.MethodReference)
+            if (semanticModel.GetOperation(cast.Expression, cancellationToken)?.Kind == OperationKind.MethodReference)
                 return false;
 
             return IsCastSafeToRemove(cast, cast.Expression, semanticModel, cancellationToken);

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
@@ -83,7 +83,8 @@ namespace Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers
 
             // Special case for converting a method group to object. The compiler issues a warning if the cast is removed:
             // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
-            if (semanticModel.GetOperation(cast.Expression, cancellationToken)?.Kind == OperationKind.MethodReference)
+            if (semanticModel.GetOperation(cast.Expression, cancellationToken)?.Kind == OperationKind.MethodReference &&
+                semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type is { SpecialType: SpecialType.System_Object })
                 return false;
 
             return IsCastSafeToRemove(cast, cast.Expression, semanticModel, cancellationToken);

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs
@@ -85,7 +85,9 @@ namespace Microsoft.CodeAnalysis.CSharp.Simplification.Simplifiers
             // warning CS8974: Converting method group 'ToString' to non-delegate type 'object'. Did you intend to invoke the method?
             if (semanticModel.GetOperation(cast.Expression, cancellationToken)?.Kind == OperationKind.MethodReference &&
                 semanticModel.GetTypeInfo(cast.Type, cancellationToken).Type is { SpecialType: SpecialType.System_Object })
+            {
                 return false;
+            }
 
             return IsCastSafeToRemove(cast, cast.Expression, semanticModel, cancellationToken);
         }

